STACKS 1.46

ACharbonneau opened the pull request:
Crossposted from issue: https://github.com/Homebrew/homebrew-science/issues/5448

All I've changed in this is the download link and sha256, and I keep getting a `fatal error: 'tr1/functional' file not found`

It looks like this issue was addressed in the previous formula update https://github.com/Homebrew/homebrew-science/pull/2265 so I'm not sure why it's back or what to do now.

I have homebrew completely updated, including the gcc compiler, and am running OSX Sierra 10.12.4 

I get the same error when I try to install on a linux machine, so it doesn't seem to be just an OSX problem anymore. 


### Have you:

- [X] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [X] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?
- [X] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [X] Built your formula locally prior to submission with `brew install <formula>`?


### Updates to existing formula: have you

- [ X] Removed the `revision` line, if any, when bumping the version number?
- [ X] Added/bumped the `revision` number if the changes affect the precompiled bottles?
- [X] Not altered the `bottle` section?
- [X] Checked that the tests still pass?

sjackman commented:
TravisCI
```
g++-5: error: ./htslib/libhts.a: No such file or directory
```
https://travis-ci.org/Homebrew/homebrew-science/builds/223748725#L1397

If it's a make parallelization issue, then either
```ruby
    system "make"
    ENV.deparallelize
    system "make", "install"
```
or
```ruby
    ENV.deparallelize
    system "make", "install"
```
may resolve it.

sjackman commented:
CircleCI
```
checking for functional regular expressions... configure: error: Regular expressions are not functional, you need g++ 4.9.0 or greater.
```
Add…
```ruby
fails_with :gcc => "4.8"
```

sjackman commented:
If you're getting undefined symbol errors on Linux, try changing
```diff
-if MacOS.version < :mavericks
+if OS.mac? && MacOS.version < :mavericks
```

sjackman commented on stacks.rb:
```diff
@@ -25,14 +25,19 @@ class Stacks < Formula
 
   needs :cxx11
+  fails_with :gcc => "4.8"
+ # needs :openmp if build.with? "openmp"
```
The syntax here is:
```ruby
  option "with-openmp", "Enable OpenMP multithreading"
  needs :openmp if build.with? "openmp"
```

sjackman commented:
Glad to see that you got it working! Good show! :tada:

Could you please drop the merge commit cafdbfdd1ce15108051f37ad22b17d9c1fe703f6 squash, the other three commits to a single commit, and rebase onto `master`?
You can use `git rebase -i origin/master` for this purpose.
Delete the merge commit. Keep the first `pick` and change the other two `pick` to `squash`.

ACharbonneau commented:
It took a couple tries to get the rebasing to work, but I think I got it. 

This recipe compiles and seems to work on Linux.

However, if I brew install with the recipe I'm still getting the `<tr1/functional>` error on OS X:
Gist:  https://gist.github.com/ACharbonneau/833c53f4e67be093a498f4d173dc0654

But, if I do the exact same install (as far as I can tell) but in interactive mode, it *does* install:
Gist: https://gist.github.com/ACharbonneau/39572f41c717a2e9ba1b36426346cd46

I have no idea what that implies or how to fix it.

sjackman commented:
Other than this `brew audit stacks` failure, it successfully built on Linuxbrew!
```
==> brew audit homebrew/science/stacks --online
Error: 1 problem in 1 formula
homebrew/science/stacks:
  * The URL http://creskolab.uoregon.edu/stacks/ could use HTTPS rather than HTTP
==> FAILED
```
https://travis-ci.org/Homebrew/homebrew-science/builds/224145100#L1296
Please change the `homepage` URL.
```diff
-homepage "http://creskolab.uoregon.edu/stacks/"
+homepage "https://creskolab.uoregon.edu/stacks/"
```

sjackman commented:
I can't replicate the build failure on my own MacOS 10.11 El Capitan box, which makes troubleshooting difficult.

sjackman commented:
The `google-sparsehash` dependency should be a `:build` dependency, but that's unrelated to your error. It's not required to be installed at run-time, because it's a header-only package (no installed libraries).
```ruby
  if OS.mac? && MacOS.version < :mavericks
    depends_on "google-sparsehash" => [:build, :recommended, "c++11"]
  else
    depends_on "google-sparsehash" => [:build, :recommended]
  end
```

sjackman commented:
I've got a fix for macOS! Add to the formula:
```ruby
  # Fix error: 'tr1/functional' file not found
  patch :DATA
```
and append to the very end of the file
```diff
__END__
diff --git a/src/DNANSeq.h b/src/DNANSeq.h
index 62b5d91..9c2ff12 100644
--- a/src/DNANSeq.h
+++ b/src/DNANSeq.h
@@ -25,7 +25,7 @@
 #include <limits.h>
 
 #include <functional> //std::hash
-#ifdef HAVE_SPARSEHASH
+#if 0
 #include <tr1/functional>
 #endif
 
@@ -115,7 +115,7 @@ struct hash<DNANSeq> {
     }
 };
 
-#ifdef HAVE_SPARSEHASH
+#if 0
 namespace tr1 {
 template<>
 struct hash<DNANSeq> {
```

I've sent an e-mail to the Stacks mailing list:
https://groups.google.com/forum/#!topic/stacks-users/b8dpBrLowXI

sjackman commented:
Please add the `:build` tag to the `google-sparsehash` dependency.

ACharbonneau commented:
Oops! I thought the audit was telling me to change the download URL, and I had checked that that wouldn't work. 

Fixed audit. Added patch. Tested on OS X and linux. Rebased again.

sjackman commented:
Merged! Thanks for your contribution to Homebrew-Science, Amanda!
